CI test_vlv_recreation_reindex fails on LMDB

**Issue Description**
A clear and concise description of what the bug is.
CI test [https://github.com/389ds/389-ds-base/actions/runs/6463318723/job/17546826993
](https://github.com/389ds/389-ds-base/actions/runs/6463318723/job/17546826993) fails on LMDB:

```
dirsrvtests/tests/suites/vlv/regression_test.py::test_vlv_recreation_reindex FAILED [100%]
...
        for i, (dn, entry) in enumerate(result, start=imin):
            assert i <= imax
            expected_dn = f'uid=testuser{i},ou=People,dc=example,dc=com'
            log.debug(f'found {dn} expected {expected_dn}')
>           assert dn.lower() == expected_dn.lower()
E           AssertionError: assert 'uid=testuser...xample,dc=com' == 'uid=testuser...xample,dc=com'
E             - uid=testuser4498,ou=people,dc=example,dc=com
E             ?              ^^^
E             + uid=testuser4501,ou=people,dc=example,dc=com
E             ?              ^^^

dirsrvtests/tests/suites/vlv/regression_test.py:64: AssertionError
```

**Package Version and Platform:**
 - Platform: Fedora
 - Package and version: main branch
 - Browser [e.g. chrome, safari]

**Steps to Reproduce**
1. See CI (LMDB Test( vlv)) report. 

**Expected results**
Test should pass.

**Additional context**
Looks like there is a bug in the mdb vlv reindexing code

--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c
@@ -2300,11 +2300,21 @@ int dbmdb_cursor_get_recno(dbi_cursor_t *cursor, MDB_val *dbmdb_key, MDB_val *db
     MDB_cursor *newcur = NULL;
     int cmpres = 0;
     int rc = 0;
+    struct ldbminfo *li = (struct ldbminfo *)cursor->be->be_database->plg_private;
+    dbmdb_ctx_t *ctx = MDB_CONFIG(li);
+    dbmdb_dbi_t *dbi = &ctx->dbi_slots[mdb_cursor_dbi(cursor->cur)];
 
     rc = MDB_CURSOR_GET(cursor->cur, &curpos_key, &curpos_data, MDB_GET_CURRENT);
     if (rc != 0) {
         return rc;
     }
+    if (dbi->dbname[0] == '~') {
+        /* Already hanfling the vlv cache (probably running dbscan) */
+        *dbmdb_key = curpos_key;
+        *dbmdb_data = curpos_data;
+        return rc;
+    }
+
     dbmdb_generate_recno_cache_key_by_data(&cache_key, &curpos_key, &curpos_data);
 
     rc = dbmdb_recno_cache_lookup(cursor, &cache_key, &rce);
@@ -2343,12 +2353,12 @@ int dbmdb_cursor_set_recno(dbi_cursor_t *cursor, MDB_val *dbmdb_key, MDB_val *db
     MDB_val cache_key = {0};
     dbi_recno_t recno;
     int rc;
-    if (!dbmdb_data || !dbmdb_data->mv_data) {
-        slapi_log_err(SLAPI_LOG_ERR, "dbmdb_cursor_set_recno", "invalid dbmdb_data parameter (should be a dbi_recno_t)\n");
+    if (!dbmdb_key || !dbmdb_key->mv_data) {
+        slapi_log_err(SLAPI_LOG_ERR, "dbmdb_cursor_set_recno", "invalid dbmdb_key parameter (should be a dbi_recno_t)\n");
         return DBI_RC_INVALID;
     }
 
-    memcpy(&recno, dbmdb_data->mv_data, sizeof (dbi_recno_t));
+    memcpy(&recno, dbmdb_key->mv_data, sizeof (dbi_recno_t));
     dbmdb_generate_recno_cache_key_by_recno(&cache_key, recno);
     rc = dbmdb_recno_cache_lookup(cursor, &cache_key, &rce);
     if (rc ==0) {

--- a/ldap/servers/slapd/back-ldbm/ldbm_search.c
+++ b/ldap/servers/slapd/back-ldbm/ldbm_search.c
@@ -157,7 +157,10 @@ ldbm_back_search_cleanup(Slapi_PBlock *pb,
     ldbm_instance *inst;
     back_search_result_set *sr = NULL;
     int free_candidates = 1;
+    Slapi_Operation *op = NULL;
 
+
+    slapi_pblock_get(pb, SLAPI_OPERATION, &op);
     slapi_pblock_get(pb, SLAPI_BACKEND, &be);
     inst = (ldbm_instance *)be->be_instance_info;
     /*
@@ -165,7 +168,13 @@ ldbm_back_search_cleanup(Slapi_PBlock *pb,
      * clean it up for the following sessions.
      */
     slapi_be_unset_flag(be, SLAPI_BE_FLAG_DONT_BYPASS_FILTERTEST);
-    CACHE_RETURN(&inst->inst_cache, &e); /* NULL e is handled correctly */
+    if (e != operation_get_target_entry(op)) {
+        /*
+         * Target entry is released later on
+         * (in cache_return_target_entry called by op_shared_search).
+         */
+        CACHE_RETURN(&inst->inst_cache, &e); /* NULL e is handled correctly */
+    }
     if (inst->inst_ref_count) {
         slapi_counter_decrement(inst->inst_ref_count);
     }

--- a/ldap/servers/slapd/back-ldbm/vlv.c
+++ b/ldap/servers/slapd/back-ldbm/vlv.c
@@ -788,7 +788,8 @@ do_vlv_update_index(back_txn *txn, struct ldbminfo *li, Slapi_PBlock *pb, struct
     } else {
         /* Very bad idea to do this outside of a transaction */
     }
-    if (priv->dblayer_clear_vlv_cache_fn) {
+    if (txn && !txn->back_special_handling_fn && priv->dblayer_clear_vlv_cache_fn) {
+        /* If there is a txn and it is not an import pseudo txn then clear the vlv cache */
         priv->dblayer_clear_vlv_cache_fn(be, db_txn, db);
     }
     data.size = sizeof(entry->ep_id);

